### PR TITLE
Add support thread-local directx (OpenCL interop) initialization

### PR DESCRIPTION
**Pull request by masa-iwm**
Add support for DirectX-OpenCL interop thread-local initialization.

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Linux x64,Linux OpenCL,Win64 OpenCL,Docs
```

### CODE REVIEW

File: modules/core/src/directx.cpp [hunk 1]
```diff
@@ -1561,26 +1659,23 @@ void convertFromD3D10Texture2D(ID3D10Texture2D* pD3D10Texture2D, OutputArray dst
...
     using namespace cv::ocl;
-    static cl_platform_id initializedPlatform = NULL;
-    cl_platform_id platform = (cl_platform_id)Platform::getDefault().ptr();
-    if (initializedPlatform != platform)
+    OpenCLDirectXImpl& impl = getImpl();
+    cl_platform_id platform = impl.getPlatform();
+    if (impl.initializedPlatform9 != platform)
     {
-        clCreateFromDX9MediaSurfaceKHR = (clCreateFromDX9MediaSurfaceKHR_fn)
+        impl.clCreateFromDX9MediaSurfaceKHR = (clCreateFromDX9MediaSurfaceKHR_fn)
```
Comment (alalek) [inline modules/core/src/directx.cpp:1670]: Perhaps such initialization should be moved to `OpenCLDirectXImpl` methods to keep code responsibilities clean.

File: modules/core/src/directx.cpp [hunk 2]
```diff
@@ -234,12 +234,116 @@ int getTypeFromD3DFORMAT(const int iD3DFORMAT)
...
+    cl_platform_id initializedPlatform9 = NULL;
+    cl_platform_id initializedPlatform10 = NULL;
+    cl_platform_id initializedPlatform11 = NULL;
+
+    cl_platform_id getPlatform()
+    {
+        if (!platform)
+        {
+            cl_device_id current_device = (cl_device_id)ocl::Device::getDefault().ptr();
+            CV_Assert(current_device);
```
Comment (alalek) [inline modules/core/src/directx.cpp:275]: Perhaps it is better to move this below (after platforms query, before `bool found = false;`)

File: modules/core/src/directx.cpp [hunk 3]
```diff
@@ -234,12 +234,116 @@ int getTypeFromD3DFORMAT(const int iD3DFORMAT)
...
+    clEnqueueReleaseDX9MediaSurfacesKHR_fn clEnqueueReleaseDX9MediaSurfacesKHR = NULL;
+
+    cl_platform_id initializedPlatform9 = NULL;
+    cl_platform_id initializedPlatform10 = NULL;
+    cl_platform_id initializedPlatform11 = NULL;
+
+    cl_platform_id getPlatform()
+    {
+        if (!platform)
+        {
```
Comment (alalek) [inline modules/core/src/directx.cpp:273]: It make sense to add this check here: `CV_Assert(cv::ocl::haveOpenCV());`

File: modules/core/src/ocl.cpp [hunk 4]
```diff
@@ -113,6 +113,14 @@
 
 #include "opencv2/core/opencl/runtime/opencl_core.hpp"
 
+#ifdef HAVE_DIRECTX
+namespace cv { namespace directx {
+struct OpenCLDirectXImpl;
+OpenCLDirectXImpl* createDirectXImpl();
+void deleteDirectXImpl(OpenCLDirectXImpl**);
```
Comment (alalek) [inline modules/core/src/ocl.cpp:120]: Please move this into `modules/core/src/directx.cpp` and use it instead (here and in directx.hpp).

Consider using this namespace:
- `cv::directx::internal`
- or `cv::directx::detail`

File: modules/core/src/directx.cpp [hunk 5]
```diff
@@ -234,12 +234,116 @@ int getTypeFromD3DFORMAT(const int iD3DFORMAT)
...
+        *p = NULL;
+    }
+}
+OpenCLDirectXImpl* getDirectXImpl(ocl::Context& ctx);
+OpenCLDirectXImpl& getImpl()
+{
+    OpenCLDirectXImpl* i = getDirectXImpl(ocl::Context::getDefault());
+    CV_Assert(i);
+    return *i;
+}
```
Comment (alalek) [inline modules/core/src/directx.cpp:342]: static call of `OpenCLDirectXImpl`?

File: modules/core/src/directx.cpp [hunk 6]
```diff
@@ -234,12 +234,116 @@ int getTypeFromD3DFORMAT(const int iD3DFORMAT)
...
+                status = clGetDeviceIDs(platforms[i], CL_DEVICE_TYPE_ALL, numDevices, &devices[0], 0);
+                if (status != CL_SUCCESS)
+                    CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get devices");
+
+                for (int j = 0; !found && j < (int)numDevices; j++)
+                {
+                    if (current_device == devices[j])
+                    {
+                        platform = platforms[i];
+                        found = true;
```
Comment (alalek) [inline modules/core/src/directx.cpp:309]: Perhaps you want to use `clGetDeviceInfo()` call with `CL_DEVICE_PLATFORM`.

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

BTW, did you try to validate this patch on configuration with multiple GPUs on some example?

File: modules/core/src/directx.cpp (hunk 3 above)
Comment (masa-iwm) [reply modules/core/src/directx.cpp:273]: Maybe haveOpenCL instead of haveOpenCV?

File: modules/core/src/directx.cpp (hunk 6 above)
Comment (masa-iwm) [reply modules/core/src/directx.cpp:309]: Oh! Thank you!
Certainly the `platform` can be obtained using `clGetDeviceInfo()`. Change it to use this.

File: modules/core/src/directx.cpp (hunk 2 above)
Comment (masa-iwm) [reply modules/core/src/directx.cpp:275]: OK, I move this below.
(I'm changing it to use `clDeviceInfo()`, so it doesn't have much effect.)

File: modules/core/src/directx.cpp (hunk 1 above)
Comment (masa-iwm) [reply modules/core/src/directx.cpp:1670]: OK, move all the initialization to `OpenCLDirectXImpl`.

File: modules/core/src/ocl.cpp (hunk 4 above)
Comment (masa-iwm) [reply modules/core/src/ocl.cpp:120]: OK, move this into `modules/core/src/directx.cpp` and using `cv::directx::internal` namespace.

**Comment by masa-iwm** [comment]
It has been verified in my multi-GPU environment.
In the software I'm developing, I need to convert DirectX 11 textures to UMat, but if the primary GPU is Intel and the secondary GPU is NVIDIA, using OpenCV 4.5 caused an error, so I created a this patch.

**Comment by masa-iwm** [comment]
The error in the above comment occurs when using an NVIDIA GPU.

**Review by alalek — APPROVED** [review]
Thank you for contribution :+1:

**Pull request merged**